File: 2/2/pdoo/prac/p1/ruby/deepspace.rb
#encoding:utf-8
module Deepspace
  # Resultados posibles de un combate
  module CombatResult
    ENEMYWINS = :enemywins
    NOCOMBAT = :nocombat
    STATIONSCAPES = :stationscapes
    STATIONWINS = :stationwins
  end
  
  # Tipos de personajes
  module GameCharacter
    ENEMYSTARSHIP = :enemystarship
    SPACESTATION = :spacestation
  end

  # Resultados posibles de un disparo
  module ShotResult
    DONOTRESIST = :donotresist
    RESIST = :resist
  end
  
  # Tipos de armas
  module WeaponType
    class Type
      attr_reader :power

      def initialize(power_val)
        @power = power_val
      end
   end

    LASER = Type.new(2.0)
    MISSILE = Type.new(3.0)
    PLSMA = Type.new(4.0)
  end

  # Botín
  class Loot
    attr_reader :nSupplies, :nWeapons, :nShields, :nHangars, :nMedals

    def initialize(supplies_num,weapons_num,shields_num,hangars_num,medals_num)
      @nSupplies = supplies_n
      @nWeapons = weapons_n
      @nShields = shields_n
      @nHangars = hangars_num
      @nMedals = medals_num
    end
  end
  
  # Paquete de suministros
  class SuppliesPackage
    attr_reader :ammoPower, :fuelUnits, :shieldPower
    
    def initialize(ammo_p,fuel_u,shield_p)
      @ammoPower = ammo_p
      @fuelUnits = fuel_u
      @shieldPower = shield_p
    end
    
    def self.newCopy(supplies_package)
      copy = SuppliesPackage.new(supplies_package.ammoPower,supplies_package.fuelUnits,supplies_package.shieldPower)
      return copy
    end
  end
  
  # Potenciadores de escudo
  class ShieldBooster
    attr_reader :name, :boost, :uses
    
    def initialize(booster_name,booster_value,booster_uses)
      @name = booster_name
      @boost = booster_value
      @uses = booster_uses
    end
   
    def self.newCopy(shield_booster)
      copy = ShieldBooster.new(shield_booster.name,shield_booster.boost,shield_booster.uses)
      return copy
    end

    def useIt
      if @uses > 1
        @uses = @uses - 1
        return @boost
      else
        return 1.0
      end
    end
  end

  # Armas de la estación
  class Weapon
    attr_reader :type, :uses

    def initialize(weapon_name,weapon_type,weapon_uses)
      @name = weapon_name
      @type = weapon_type
      @uses = weapon_uses
    end

    def self.newCopy(weapon)
      copy = Weapon.new(weapon.name,weapon.type,weapon.uses)
      return copy
    end
  
    def power
      @type.power
    end

    def useIt
      if @uses > 1
        @uses = @uses - 1
        return power
      else
        return 1.0
      end
    end
  end
end

File: 2/2/pdoo/prac/p1/java/deepspace/src/deepspace/deepspace.java
/**
 * @author Manuel Gachs Ballegeer
 */
package deepspace;
enum CombatResult {ENEMYWINS, NOCOMBAT, STATIONESCAPES, STATIONWINS} ///< Posibles resultados del combate
enum GameCharacter {ENEMYSTARSHIP, SPACESTATION} ///< Tipos de personajes
enum ShotResult {DONOTRESIST, RESIST} ///< Resultados posibles de un disparo
/**
 * @brief Tipos de armas
 */
enum WeaponType {
    LASER(2.0f),
    MISSILE(3.0f),
    PLASMA(4.0f);
    private float power;
    WeaponType(float powerValue) {power = powerValue;}
    float getPower() {return power;}
}
/**
 *  @brief Botín al vencer a una nave enemiga
 */
class Loot {
    private int nSupplies;
    private int nWeapons;
    private int nShields;
    private int nHangars;
    private int nMedals;
    
    Loot(int suppliesNumber,int weaponsNumber,int shieldsNumber,int hangarsNumber,int medalsNumber) {
        nSupplies = suppliesNumber;
        nWeapons = weaponsNumber;
        nShields = shieldsNumber;
        nHangars = hangarsNumber;
        nMedals = medalsNumber;
    }
    public int getNSupplies() {return nSupplies;}
    public int getNWeapons() {return nWeapons;}
    public int getNShields() {return nShields;}
    public int getNHangars() {return nHangars;}
    public int getNMedals() {return nMedals;}
}
/**
 *  @brief Paquete de suministros
 */
class SuppliesPackage {
    private float ammoPower;
    private float fuelUnits;
    private float shieldPower;
    
    SuppliesPackage(float powerAmmo,float unitsFuel,float powerShield) {
        ammoPower = powerAmmo;
        fuelUnits = unitsFuel;
        shieldPower = powerShield;
    }
    SuppliesPackage(SuppliesPackage s) {
        ammoPower = s.ammoPower;
        fuelUnits = s.fuelUnits;
        shieldPower = s.shieldPower;
    }
    public float getAmmoPower() {return ammoPower;}
    public float getFuelUnits() {return fuelUnits;}
    public float getShieldPower() {return shieldPower;}        
}
/**
 * @brief Potenciadores de escudo
 */
class ShieldBooster {
    private String name;
    private float boost;
    private int uses;
    
    ShieldBooster(String BoosterName,float BoostValue,int BoosterUses) {
        name = BoosterName;
        boost = BoostValue;
        uses = BoosterUses;
    }
    ShieldBooster(ShieldBooster s) {
        name = s.name;
        boost = s.boost;
        uses = s.uses;
    }
    public float getBoost() {return boost;}
    public int getUses() {return uses;}
    public float useIt() {
        if (uses>0) {
            uses = uses - 1;
            return boost;
        } else
            return 1.0f;
    }
}
/**
 * @brief Armas de las que dispone una estación
 */
class Weapon {
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String weaponName,WeaponType weaponType,int weaponUses) {
        name = weaponName;
        type = weaponType;
        uses = weaponUses;
    }
    Weapon(Weapon w) {
        name = w.name;
        type = w.type;
        uses = w.uses;
    }
    public WeaponType getType() {return type;}
    public int getUses() {return uses;}
    public float power() {return type.getPower();}
    public float useIt() {
        if (uses>0) {
            uses = uses - 1;
            return this.power();
        } else
            return 1.0f;
    }
}